packages/next/build/polyfills/process.js

```diff
@@ -1 +1,4 @@
-module.exports = global.process || require('../../compiled/process')
+module.exports =
+  typeof global.process?.env === 'object'
+    ? global.process
+    : require('../../compiled/process')
```

Comment:
I think this needs to mutate `global.process` if it exists because its referencing the DOM element so you just need to add `process.env` to it too.

Reply:
I don't think that's a good idea. I think we should not use the global `process` object when it's not a "process" thing.

Otherwise you get random properties from the DOM objects into the process.

Reply:
Sounds like a production decision, I'm fine either way. I'm more biased towards the current implementation, because the following code:

```jsx
// pages/my-page.js
import {useEffect} from 'react'
export default function MyComponent() {
  useEffect(() => {
    console.log(window.process.dataset, process.dataset)
  });
  return <div id="preview" data-hey="ho">Hello, {process.env.NEXT_PUBLIC_VISITOR}</div>
}
```

renders `Hello, stranger` and console logs `DOMStringMap { hey: "ho" }, undefined`. So we don't "leak" DOM data into what is treated as Node.js API when server rendering. The `window.` usage is explicit.